lib/src/test/java/io/ably/lib/test/rest/RestJWTTest.java

```diff
@@ -0,0 +1,157 @@
+	Param[] tokenEmbedded = new Param[]{ new Param("jwtType", "embedded") };
+	Param[] tokenEmbeddedAndEncrypted = new Param[]{ new Param("jwtType", "embedded"), new Param("encrypted", 1) };
+	Param[] jwtReturnType = new Param[]{ new Param("returnType", "jwt") };
+
+	@Before
+	public void setUpBefore() throws Exception {
+		jwtRequesterOptions = createOptions(testVars.keys[0].keyStr);
+		jwtRequesterOptions.authUrl = echoServer;
+		options = new ClientOptions();
+		options.environment = "sandbox";
+	}
+
+	/**
+	 * Base request of a JWT token (RSA8g RSA8c)
+	 */
+	@Test
+	public void authjwtrequest() {
+		try {
+			options.token = getTokenWithParams(validKeys);
+			AblyRest client = new AblyRest(options);
+			PaginatedResult<Stats> stats = client.stats(null);
+			assertNotNull("Stats should not be null", stats);
+		} catch (AblyException e) {
+			e.printStackTrace();
+			fail("authjwtrequest: Unexpected exception");
+		}
+	}
+
+	/**
+	 * Base request of a JWT token with wrong credentials (RSA8g RSA8c)
+	 */
+	@Test
+	public void authjwtrequestwrongkeys() {
+		try {
+			options.token = getTokenWithParams(invalidKeys);
+			AblyRest client = new AblyRest(options);
+			PaginatedResult<Stats> stats = client.stats(null);
+		} catch (AblyException e) {
+			assertEquals("Unexpected code from exception", 40144, e.errorInfo.code);
+			assertEquals("Unexpected statusCode from exception", 401, e.errorInfo.statusCode);
+			assertEquals("Error message not matching the expected one", "Error verifying JWT; err = invalid signature", e.errorInfo.message);
+		}
+	}
+
+	/**
+	 * Request of a JWT token that embeds and Ably token (RSC1 RSC1a RSC1c RSA3d)
+	 */
+	@Test
+	public void authjwtrequestembeddedtoken() {
+		try {
+			options.token = getTokenWithParams(mergeParams(validKeys, tokenEmbedded));
+			AblyRest client = new AblyRest(options);
+			PaginatedResult<Stats> stats = client.stats(null);
+			assertNotNull("Stats should not be null", stats);
+		} catch (AblyException e) {
+			e.printStackTrace();
+			fail("authjwtrequestembeddedtoken: Unexpected exception");
+		}
+	}
+
+	/**
+	 * Request of a JWT token that embeds and Ably token and is encrypted (RSC1 RSC1a RSC1c RSA3d)
+	 */
+	@Test
+	public void authjwtrequestembeddedtokenencrypted() {
+		try {
+			options.token = getTokenWithParams(mergeParams(validKeys, tokenEmbeddedAndEncrypted));
+			AblyRest client = new AblyRest(options);
+			PaginatedResult<Stats> stats = client.stats(null);
+			assertNotNull("Stats should not be null", stats);
+		} catch (AblyException e) {
+			e.printStackTrace();
+			fail("authjwtrequestembeddedtokenencrypted: Unexpected exception");
+		}
+	}
+
+	/**
+	 * Request of a JWT token that is returned with application/jwt content type (RSA4f, RSA8c)
+	 */
+	@Test
+	public void authjwtrequestreturntype() {
+		try {
+			options.token = getTokenWithParams(mergeParams(validKeys, jwtReturnType));
```

Comment:
I thought `getTokenWithParams()` was a helper method to obtain a token, but the AblyRest instance created in the test itself was the instance that's the subject of the test. If that's the case, then this test should be using `authURL` in this instance, and asking the echoserver to reply with `application/jwt` here, rather than doing that in `getTokenWithParams()`.

Reply:
done https://github.com/ably/ably-java/pull/393/commits/6a02ea7486214c7a0b1295ff8ad8dbbb277e3f36